### ISSUE TITLE
Rivian Phone Key not found

### ISSUE THREAD

**Issue by maddox** [bretterer/home-assistant-rivian#181]
I'm attempting to pair a new R1S with the latest software and can't seem to get it to work. It keeps failing with `Rivian Phone Key not found`.

I was able to add my truck to the integration. I was able to get into the pairing stage of the connection. The truck saw my home assistant as a key, but it never successfully pairs. The truck keeps saying "Let's try again". And the logs show:

    2024-11-20 19:39:23.805 ERROR (MainThread) [custom_components.rivian.button] Rivian Phone Key not found

## Setup

1. Home Assistant running in docker
2. home-assistant-rivian added via HACS
3. using an esp32 with bluetooth-proxy

## Steps

1. ✅ Login with Rivian account
2. ✅ Truck sensors show up fine
3. ✅ Configure > Enable vehicle control
4. ✅ Truck sees the new Phone Key named after my Home Assistant instance
5. ✅ Tap Setup in truck, and click Pair in Home Assistant
6. ❌ Pairing times out without being successful. Logs report what was pasted above.

**Comment by maddox** [comment]
Looks like this is because it's a 2025, which is not available for pairing yet due to updated bluetooth hardware.

**Comment by mporet** [comment]
Is this feature planned?

**Comment by tmack8001** [comment]
@mporet that depends. Are you offering either a) to be a contributor or b) to loan your 2025 truck to an existing contributor?

Unfortunately non of us have a 2025 build truck and there are changes to the Bluetooth stack we don't have hands on visibility to 😕 so it makes it hard to build compatible software with something that isn't documented if we don't have access to one ourselves

**Comment by tmack8001** [comment]
Goes for anyone that comes reporting this issue. Until a developer on the project have a 2025 in hand or someone comes willing to develop this themselves for this project it is going to be difficult.

What would be helpful is to log BT traces between mobile device PaaK and the truck so some reverse engineering can take place.

**Comment by mporet** [comment]
I lack the experience with HASS development to be a direct contributor, but I do have a 2025 vehicle and would be wiling to gather data, where helpful

**Comment by maddox** [comment]
I’m willing to gather any info as well. I am a developer and am familiar with esp32. I’m just not experienced in the hardware protocols like this takes. 

But I can take directions well, operate a test rig, and can report back with great comms!

**Comment by mattbrad2** [comment]
I have a 2025 Tri-Max. I have a programming background, but haven't done much with developing for HACS. I would certainly be willing to contribute anything I can to help with this.

**Comment by natekspencer** [comment]
You can look at https://github.com/bretterer/rivian-python-client/blob/main/src/rivian/ble.py for how we currently handle gen 1 BLE pairing as a start. Outside of that, we need BLE packet captures to then try to reverse engineer the flow. Ideally, someone with a gen 2 would be able to reproduce the entire process a few times so that we could then get it into HA.

**Comment by maddox** [comment]
If there's any direction that can be made for setting up an esp32 to capture the BLE packets, I'd be more than happy to set something up and try to log as much as I can.

**Comment by yetdog** [comment]
> If there's any direction that can be made for setting up an esp32 to capture the BLE packets, I'd be more than happy to set something up and try to log as much as I can.

SImilar. most searches for BLE sniffer go to device detectors and not packet analyzers. Then I stumbled upon $20,000 (!!) pre-made options. I can play with ESP32s as well but I'm not seeing much by the way of that out there already.

New Gen2 owner and I'd love to throw in on this.

**Comment by frankrosano** [comment]
Hi @natekspencer - I used a Pixel 7a I had laying around, enabled the Bluetooth HCI snoop log, then paired the device as a phone key on our Gen2 R1S. I captured the log and am sharing it here:

[btsnoop_hci.log](https://github.com/user-attachments/files/19653111/btsnoop_hci.log)

Let me know if this is useful, or if further pairing attempts are required.

**Comment by runxmedia** [comment]
THANK YOU!! I tried SO many times to get to my HCI snoop logs on my samsung and I COULD NOT get to them without rooting

**Comment by igillespie** [comment]
I wonder if anyone has had a chance to look at the log @frankrosano posted. I do have access to a Gen 2 R1S and I've tried pairing with some code I wrote myself, but it didn't work (I don't know Bluetooth technology very well at all).

**Comment by yetdog** [comment]
Any chance to peek at the dump yet, @natekspencer ?

**Comment by natekspencer** [comment]
Unfortunately, all I have done is glanced at the file. The biggest issue is that I don't have a gen2 vehicle to test and see if it's reacting appropriately. The process to connect the gen1 was already tedious and time consuming. As I said before, ideally someone with a gen2 could take this log, program something in python, validate, and then open up a PR with the appropriate changes.

**Comment by hessnd** [comment]
Hey @natekspencer, I have a gen2 R1T and would be interested in collaborating.

**Comment by 7909linder** [comment]
> Whats the status on this? I have a gen2 r1s (currently in the shop but hopefully out soon) and am willing to help. I'll also give away a coA³st cargo slide for the R1S to whoever cracks this.

Any update ?

**Comment by mporet** [comment]
You might try the Rivian discord -  one or two devs are known to hang out
there

**Comment by sundar2012** [comment]
> You might try the Rivian discord -  one or two devs are known to hang out
> there
> […](#)

can you drop a link to the discord channel?  thanks!

**Comment by mporet** [comment]
https://discord.gg/rivian-892997443896111105

**Comment by spyder94** [comment]
> I've done some some vibe coding to do this, but don't have a Gen 2 to test. [jrgutier/rivian-python-client@c10460c](https://github.com/jrgutier/rivian-python-client/commit/c10460c1c8f7663a418e6fa9a4e6cff985ae1e88)

Is there anything different that needs to be done to use this? More than happy to give it a try.

**Comment by spyder94** [comment]
@mattbrad2 , I tried removing this integration and unloaded it from HACS. Then added @jrgutier 's version. When trying to add the integration, it fails almost immediately with a 500 error. Doesn't look I can add an issue to his fork either. Anybody else able to get any where with it?

**Comment by maddox** [comment]
I'm pretty sure that implementation was 100% vibe coded and has never been confirmed to have ever worked. The contributor even mentioned that they didn't even own a Gen 2 vehicle.

**Comment by mattbrad2** [comment]
> I'm pretty sure that implementation was 100% vibe coded and has never been confirmed to have ever worked. The contributor even mentioned that they didn't even own a Gen 2 vehicle.

Yep, but it appears he's updated it several times since then. I just assumed it was *working*, but wasn't sure about Gen2 pairing. Since none of the developers have a Gen2 vehicle, I'm not sure what else we can do.

**Comment by maddox** [comment]
Same. If anyone could design a debug rig for us to collect info that would help build it out, I'd be HAPPY to utilize it. But I just have zero experience in this field (wireless, etc).

**Comment by Vasa211** [comment]
> You can look at https://github.com/bretterer/rivian-python-client/blob/main/src/rivian/ble.py for how we currently handle gen 1 BLE pairing as a start. Outside of that, we need BLE packet captures to then try to reverse engineer the flow. Ideally, someone with a gen 2 would be able to reproduce the entire process a few times so that we could then get it into HA.

Hi. I analyzed the Android app, extracted the necessary commands, etc.
I managed to connect to the car and register a key via the API — it appeared in the vehicle.
Right now I’m stuck at the pairing stage (the one with the PIN code): I can’t complete the pairing yet, but I think I’ll figure it out soon.

Is there any community where we could discuss adding Gen2 support to the add-on more actively?

I used raspberry pi for tests now

**Comment by heyhewmike** [comment]
I have a Gen 2 and willing to test. I did have a digital key appear in my Vehicle settings and had been asked to pair my key. When I go through the pairing process it asks me to check my device to begin the pairing but I am unsure as to where to approve/acknowledge the pairing request.
